File: AudioGeneratorDesktop/src/main/java/com/cesar31/audiogenerator/control/Listener.java
package com.cesar31.audiogenerator.control;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;

/**
 *
 * @author csart
 */
public class Listener extends Thread {

    private int portServer;
    private String ipClient;
    private int portClient;

    private RequestParserHandler parser;

    private Listener() {
        this.parser = new RequestParserHandler();
    }

    /**
     * Constructor de listener
     *
     * @param portServer puerto donde el servidor escucha peticiones
     * @param ipClient direccion ip del cliente
     * @param portClient puerto donde el cliente escucha peticiones
     */
    public Listener(int portServer, String ipClient, int portClient) {
        this();
        this.portServer = portServer;
        this.ipClient = ipClient;
        this.portClient = portClient;
    }

    @Override
    public void run() {
        try {
            ServerSocket server = new ServerSocket(portServer);
            System.out.println("Escuhando en: " + portServer);
            while (true) {
                try ( Socket socket = server.accept()) {
                    InputStreamReader reader = new InputStreamReader(socket.getInputStream());
                    BufferedReader br = new BufferedReader(reader);
                    String message = "";
                    String line = br.readLine();
                    while (line != null) {
                        message += line;
                        line = br.readLine();
                        if (line != null) {
                            message += "\n";
                        }
                    }
                    System.out.println("Recibiendo : \n" + message);
                    // Analizar mensaje aqui
                    String response = parser.parseSource(message);
                    sendResponse(response);
                }
            }
        } catch (IOException ex) {
            ex.printStackTrace(System.out);
        }
    }

    /**
     * Enviar respuesta a cliente
     *
     * @param response mensaje que se envia al cliente
     */
    private void sendResponse(String response) {
        System.out.println("Enviado: \n" + response);
        try {
            try ( Socket socket = new Socket(ipClient, portClient); PrintWriter printWriter = new PrintWriter(socket.getOutputStream())) {
                printWriter.write(response);
                printWriter.flush();
            }
        } catch (IOException ex) {
            ex.printStackTrace(System.out);
        }
    }
}

File: AudioGeneratorClient/app/src/main/java/com/cesar31/audiogeneratorclient/task/Listener.kt
package com.cesar31.audiogeneratorclient.task

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.os.Bundle
import com.cesar31.audiogeneratorclient.ResponseActivity
import com.cesar31.audiogeneratorclient.control.ResponseParserHandler
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.Serializable
import java.net.ServerSocket

class Listener(var port: Int, var context: Context) : Thread(), Serializable {

    @SuppressLint("StaticFieldLeak")
    companion object {
        lateinit var listen: Listener
        var input: String = ""
        var response: String = ""
        fun setListener(l : Listener) {
            this.listen = l
        }

        fun getListener(): Listener {
            return this.listen
        }
    }

    override fun run() = try {
        val server = ServerSocket(port)
        while(true) {
            server.accept().use { socket ->
                val reader = InputStreamReader(socket.getInputStream())
                val br = BufferedReader(reader)
                val message = br.readText();
                // println("Recibiendo:\n$message")
                parseResponse(message)
            }
        }
    } catch(e: IOException) {
        e.printStackTrace(System.out)
    }

    private fun parseResponse(input: String) {
        response = input
        val parser = ResponseParserHandler()
        val bundle: Bundle = parser.parseResponse(input);
        getResponseActivity(bundle)
    }

    private fun getResponseActivity(bundle: Bundle) {
        val i = Intent(context, ResponseActivity::class.java)
        i.putExtras(bundle)
        context.startActivity(i)
    }
}